### PR TITLE
Parallax scrolling on paywall (iOS)

### LINKED ISSUE #1228: Mobile Paywall allows for Scrolling
## Description
On mobile (iOS chrome) the paywall allows for unpurchased content to be scrolled

## Reproduce
visit following site on mobile (iOS chrome) http://projects.mombartz.com/unlock or https://www.ouvre-boite.com/members

## Expected behavior
should not allow content to be scrolled

## Device
**Smartphone**
 - Device: iphone x
 - OS: iOS 12.1.2
 - Browser: chrome

### PR DESCRIPTION
**Pull request by cnasc**
# Description

This PR fixes #1228. It adds some logic and communication between the host page and the paywall to allow the paywall to grow as a user scrolls, thus obscuring the page content more and more as the user scrolls further and further down the page. Currently this only affects iOS, since only iOS allows users to scroll when the paywall is up. A future PR may adjust the way the paywall works so the parallax happens on other devices as well.

Comments on this implementation are most appreciated, there are a few things about it that I'm not 100% sure on. Another future PR or several will probably go to refactoring the mobile paywall experience to ensure that everything fits the screen nicely in the default state and that elements do not jump around too much when the paywall changes size.

grow
(gif is choppy for space reasons, it's actually smooth)

Overlay continues to display as before in other cases
screen shot 2019-01-30 at 10 53 54 am
screen shot 2019-01-30 at 10 53 38 am
screen shot 2019-01-30 at 10 52 30 am

# Checklist:

- [x] 1 PR, 1 purpose: my Pull Request applies to a single purpose
  - [ ] This PR only contains configuration changes (package.json, etc.)
  - [x] This PR only contains code changes (if configuration changes are required, do a separate PR first, then re-base)
- [x] My code follows the style guidelines of this project, including naming conventions
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If my code adds or changes components, I have written corresponding stories with Storybook
- [x] I have performed a self-review of my own code
- [x] If my code involves visual changes, I am adding applicable screenshots to this thread

PS: [Read how to write the perfect pull request](https://blog.github.com/2015-01-21-how-to-write-the-perfect-pull-request/)

### CODE REVIEW

File: unlock-app/src/pages/paywall.js [hunk 1]
```diff
@@ -8,17 +8,42 @@ import { LOCK_PATH_NAME_REGEXP } from '../constants'
...
+  }
+  componentDidMount() {
+    window.addEventListener('message', event => {
+      if (event.data.scrollPosition) {
+        this.handleScrollPosition(event.data.scrollPosition)
+      }
+    })
+  }
+  handleScrollPosition(scrollPosition) {
+    this.setState({ scrollPosition })
```
Comment (cellog) [inline unlock-app/src/pages/paywall.js]: I think it's slightly more performant to use a callback
```suggestion
    this.setState(state => {
      if (state.scrollPosition === scrollPosition) return null
      return { scrollPosition }
    })
```

**Comment by cellog** [comment]
nice. Can you add screenshots from the Overlay, Paywall and Demo stories to verify they are still working?

**Comment by cnasc** [comment]
> nice. Can you add screenshots from the Overlay, Paywall and Demo stories to verify they are still working?

Will take those now. Note that there is an issue with the prices in the locks that exists in master, so there will be a bit of ugliness on the paywall.

**Comment by cellog** [comment]
yes i saw that :)

**Comment by cnasc** [comment]
...and in fact, the demo and paywall stories are broken on master so nothing renders on them.

File: unlock-app/src/paywall-builder/build.js [hunk 2]
```diff
@@ -9,9 +20,30 @@ export default function buildPaywall(window, document, lockAddress, blocker) {
...
+    const viewportHeight = window.innerHeight
+    const pageHeight = document.documentElement.scrollHeight
+    const maximumScroll = pageHeight - viewportHeight
+    // Avoiding a divide-by-zero error when the page does not scroll!
+    if (maximumScroll === 0) {
+      return
+    }
+
+    const scrollPosition = baseBannerHeight() + 100 * (pageTop / maximumScroll)
+    iframe.contentWindow.postMessage({ scrollPosition }, '*')
```
Comment (cellog) [inline unlock-app/src/paywall-builder/build.js]: just as a quick note, using `'*'` is a potential security hole, but I am going to handle this in a separate PR that combines all the `postMessage` stuff into a middleware, so this is the best solution for this PR.

**Pull request merged**